core/templates/dev/head/pages/exploration_editor/translation_tab/TranslationRecordingService.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the 'License');
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS-IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for handling microphone data and mp3 audio processing.
+ */
+oppia.factory('TranslationRecordingService', ['$q', '$window',
+  function($q, $window) {
+    var AudioContext = null,
+      defer = null,
+      definedAudioContext = null, // will be defined audio context
+      isAvailable = null,
+      isRecording = false,
+      microphone = null,
+      microphoneStream = null,
+      mp3Worker = null,
+      processor = null;
+
+
+    var _initWorker = function() {
+      if (!$window.Worker) {
+        console.warn('Worker API not supported in this browser.');
+        return;
+      }
+      if (mp3Worker === null) {
+        var url = '/third_party/static/lamejs-1.2.0/' +
+          'worker-example/worker-realtime.js';
+        // config the mp3 encoding worker
+        const config = {sampleRate: 44100, bitRate: 128};
+        mp3Worker = new Worker(url);
+        mp3Worker.onmessage = function(e) {
+          switch (e.data.cmd) {
+            case 'end':
+              // async data flow
+              defer.resolve(e.data.buf);
+              break;
+            default :
+              console.warn('Unexpected command and data type from worker.');
+              // console.warn(e);
+          }
+        };
+        mp3Worker.postMessage({cmd: 'init', config: config});
+      }
+    };
+
+    var _postMessage = function(buffer) {
+      mp3Worker.postMessage({cmd: 'encode', buf: buffer});
+    };
+
+    var _stopWorker = function() {
+      if (mp3Worker) {
+        mp3Worker.terminate();
+        console.warn('Ending mp3 worker');
+        mp3Worker = null;
+      }
+    };
+
+    var _initRecorder = function() {
+      // promise required because angular is async with worker
+      AudioContext = $window.AudioContext || $window.webkitAudioContext;
+      if (AudioContext) {
+        defer = $q.defer();
+        isAvailable = true;
```

Comment:
We try to only start with is/are for function names, since in general function names start with verbs to help distinguish them from the variable, so maybe AudioContextAvailable.

Reply:
ok

Reply:
Done.

Reply:
Done.